File: C++/8/BigramDyn.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package javaapplication3;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Locale;

import java.util.Scanner;

/**
 *
 * @author ORHAN
 * @param <T>
 */
@SuppressWarnings("unchecked")
public class BigramDyn <T> implements Bigram<T> {
    
    private int dataType;
    private triple <T>  AllBigr;
    private triple <T> MapBgr;
    private int allIndex=0;
    private int mapIndex=0;
    private int capacity=0;//NUMBER OF ELEMENTS IN A FILE
    
    public BigramDyn(int data) {///CONSTRUCTOR
        
        
        dataType=data;
    }

    /**
     *
     * @param filename
     * @throws myException
     */
    @Override
    public void readFile(String filename) throws myException {
        
   
        
        try {
            Scanner scan;
            scan = new Scanner(new File(filename));
            while(scan.hasNext()){
                String next = scan.next();
                capacity+=1;
            }
        } catch (FileNotFoundException ex) {
            throw new myException("Dosya bulunamadı");
        }
        
        AllBigr=new triple<>(capacity);
        
        MapBgr=new triple<>(capacity);
        try {
            Scanner scan;
            scan = new Scanner(new File(filename));
            scan.useLocale(Locale.US);
            
            if(dataType==1){
                T token1 = null;
                T token2;
                if(scan.hasNextInt())
                    token1=(T) (Object)  scan.nextInt();
                    
                while(scan.hasNextInt()){
                    token2=(T) (Object) scan.nextInt();
                    tripleAll(token1,token2,0);
                    tripleMap(token1,token2,0);
                    
                    token1=token2;
                }
                if(scan.hasNext()){
                    throw new myException("Element integer degil");
                }
                
                
            }
            else if(dataType==2){
                T token1=(T) (Object) "";
                T token2=(T) (Object) "";
                if(scan.hasNext())
                    token1=(T) (Object) scan.next();
                while(scan.hasNext()){
                    token2=(T) (Object) scan.next();
                    
                    tripleAll(token1,token2,0);
                    tripleMap( token1,token2,0);
                    token1=token2;
                }
                /* if(scan.hasNext()){
                
                }*/
            }
            else if(dataType==3){
                T token1=(T) (Object) 0;
                T token2=(T) (Object) 0;
                if(scan.hasNextDouble()){
                    token1=(T) (Object) scan.nextDouble();
                }
                while(scan.hasNextDouble()){
                    token2=(T) (Object) scan.nextDouble();
                    
                    tripleAll(token1,token2,0);
                    tripleMap(token1,token2,0);
                    token1=token2;
                }
                if(scan.hasNext()){
                    throw new myException("Element double degil");
                }
                        
                        }
        } catch (FileNotFoundException ex) {
            throw new myException("Dosya bulunamadı");
        }
        
        
        
        
        
    }

    @Override
    public int numGrams() {
        return capacity-1;
    }

    @Override
    public int numOfGrams(T x, T y) {
        int count=0;
        for(int i=0;i<allIndex;++i){
            if(AllBigr.first[i]==x && AllBigr.second[i]==y)
                count+=1;
        }
        return count;
    }
    
    
    @Override
    public String toString(){
        String AllLine="";
        
        for(int i=0;i<mapIndex;++i)////////////////MAP TUTAN ARRAY
        {   MapBgr.value[i]=0;
            for(int j=0;j<allIndex;++j)///BUTUN BIGRAMLARI TUTAN ARRAY
            {

                if((AllBigr.first[j]==MapBgr.first[i]) && (AllBigr.second[j]==MapBgr.second[i]))
                {/////////////////////////////////////////////////////////EGER MAP TUTAN ARRAYDEKI BIGRAMDAN VAR ISE
                    MapBgr.value[i]+=1;////MAP TUTAN ARRAYDEKI VALUE KISMI ARTTIRILIR
                }
            }
        }
	
        
        
        
        for(int i=1;i<mapIndex;++i){///////////////////BUBBLE SORT
		for(int j=0;j<mapIndex-i;++j){
			
			if(MapBgr.value[j]<MapBgr.value[j+1]){
				
				triple<T> temp;
                                temp = new triple<>(1);
                                
				temp.first[0]=MapBgr.first[j];
                                temp.second[0]=MapBgr.second[j];
                                temp.value[0]=MapBgr.value[j];
                                
				MapBgr.first[j]=MapBgr.first[j+1];
                                MapBgr.second[j]=MapBgr.second[j+1];
                                MapBgr.value[j]=MapBgr.value[j+1];
                                
				MapBgr.first[j+1]=temp.first[0];
			        MapBgr.second[j+1]=temp.second[0];
                                MapBgr.value[j+1]=temp.value[0];
                        }
		}
	}///////////////////////////////////////////////////////
        for(int i=0;i<mapIndex;++i){
            AllLine+="(";
            AllLine+=String.valueOf( MapBgr.first[i]);
            AllLine+=",";
            AllLine+=String.valueOf(MapBgr.second[i]);
            AllLine+=")";
            AllLine+="=";
            AllLine+=String.valueOf(MapBgr.value[i]);
            AllLine+=System.lineSeparator();
        }
        return AllLine;
    }
    /**
     *
     * @param first
     * @param second
     * @param val
     */
    public void tripleAll(T  first,T second,int val){
       
        AllBigr.first[allIndex]= first;
        AllBigr.second[allIndex]=second;
        AllBigr.value[allIndex]= val;
        allIndex+=1;
     
    }
    public void tripleMap(T first,T second,int val){
        
        for(int i=0;i<=mapIndex;++i){
            if(MapBgr.first[i]==first && MapBgr.second[i]==second)
            {
                return;
            }
        }
        MapBgr.first[mapIndex]=(T) first;
        MapBgr.second[mapIndex]=(T) second;
        MapBgr.value[mapIndex]=val;
        mapIndex+=1;
    }
    
    
    
}

File: C++/7/BigramDyn.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */








#include <iostream>
#include "BigramDyn.h"
#include <fstream>

#include <stdlib.h>


using namespace std;

template <class T>

BigramDyn<T>::BigramDyn(int data){
	dataType=data;
	
	pairAllPtr=new pairClass<T>[10];////////
	pairMapPtr=new pairClass<T>[10];/////
	pairSize=10;
	pairIndex=0;
	pairMapSize=10;
	pairMapIndex=0;
}
template <class T>
BigramDyn<T>::BigramDyn(){
	dataType=1;
	
	pairAllPtr=new pairClass<T>[10];
	pairMapPtr=new pairClass<T>[10];
	pairSize=10;
	pairIndex=0;
	pairMapSize=10;
	pairMapIndex=0;

}
template <class T>
BigramDyn<T>::~BigramDyn(){////DESTRUCTOR

	delete [] pairMapPtr;
	delete [] pairAllPtr;
	

}

template <class T>
BigramDyn<T>::BigramDyn(const BigramDyn<T>& other){////COPY CONSTRUCTOR

	
		
		pairSize=other.pairSize;
		pairIndex=0;
		pairMapSize=other.pairMapSize;
		pairMapIndex=0;
		for(int i=0;i<pairSize;++i){
			pairAll(other.pairAllPtr[i].getLeft(),other.pairAllPtr[i].getRight(),other.pairAllPtr[i].getValue());
		}
		for(int i=0;i<pairMapSize;++i){
			pairAll(other.pairMapPtr[i].getLeft(),other.pairMapPtr[i].getRight(),other.pairMapPtr[i].getValue());
		}




}


template <class T>
const BigramDyn<T>& BigramDyn<T>::operator=(const BigramDyn<T>& other){////ASSIGNMENT OVERLOAD
	
	if(this!=&other){
		delete [] pairMapPtr;
		delete [] pairAllPtr;
		pairSize=other.pairSize;
		pairIndex=other.pairIndex;
		pairMapSize=other.pairMapSize;
		pairMapIndex=other.pairMapIndex;
		for(int i=0;i<pairIndex;++i){
			pairAllPtr[i]=other.pairAllPtr[i];
		}
		for(int i=0;i<pairMapIndex;++i){
			pairMapPtr[i]=other.pairMapPtr[i];
		}


	}
	return *this;

}






template <class T>
void BigramDyn<T>::readFile( string filename){
	T token1;
    
    T token2;

    
    ifstream dosya;
    
    dosya.open(filename);
	if(!dosya.is_open()){
		
		throw myException("File not opened");
	}
{
	string temp;
	getline(dosya,temp);
	bool flag=false;
	if(temp.size()==0 ){

		throw myException("Empty File");
	}
	for(int i=0;i<temp.size();++i)
	{
		if(temp[i]==' ')
			continue;
		else{
			flag=true;
			break;
			}
	}
	if(!flag){
	
			throw myException("Empty File");	
	}
	
}	
	dosya.close();
	dosya.open(filename);
	if(!(dosya>>token1)){
	
		throw myException("inconsistent datatype");
	}
	for(int i=0;dosya.good();++i){//////////////BU KISMA KADAR BIGRAMMAP ILE AYNI
        if(dosya>>token2){
			pairAll(token1,token2,0);////BUTUN CIFTLER TUTULUR
            pairMap(token1,token2,0);///AYNI OLAN CIFTLER TUTULMAZ
            token1=token2;
        }
		
        else if(!dosya.eof()){
			
            throw myException("inconsistent datatype");
        }
		
    }


}





template <class T>


void BigramDyn<T>::pairAll(T a1,T a2,int a3){
	
	if(pairIndex>=pairSize){////////////////EGER ARRAY BOYUTU GECILIR ISE 
		pairClass<T> temp[pairSize];
		for(int i=0;i<pairSize;++i){
			temp[i]=pairAllPtr[i];/////////////GECICI CLASS ARRAY INE DEGERLER ATILIR
				
		}
		pairSize+=10;//BOYUT ARTTIRILIR
		delete [] pairAllPtr;////DEALLOCATE
		pairAllPtr=new pairClass<T>[pairSize];///ALLOCATE
		
		for(int i=0;i<pairSize-10;++i){///DEGERLER YENI ARRAY E DOLDURULUR
			pairAllPtr[i]=temp[i];
				
		}
	
	}

	pairAllPtr[pairIndex]=pairClass<T>(a1,a2,a3);////YENI DEGER ATAMSI
	pairIndex+=1;///INDEX GUNCELLENIR



}

template <class T>


void BigramDyn<T>::pairMap(T a1,T a2,int a3){/////paırAll FONKSIYONU ILE AYNI
	
	if(pairMapIndex>=pairMapSize){
			pairClass<T> temp[pairMapSize];
			for(int i=0;i<pairMapSize;++i){
				temp[i]=pairMapPtr[i];
				
			}
			pairMapSize+=10;
			delete [] pairMapPtr;
			pairMapPtr=new pairClass<T>[pairMapSize];
			for(int i=0;i<pairMapSize-10;++i){
						pairMapPtr[i]=temp[i];
				
		}
	
	}

	for(int i=0;i<pairMapIndex;++i){///////EGER YENI BIGRAM ZATEN BIZDE VAR ISE ATAMA YAPMAZ
		if(a1==pairMapPtr[i].getLeft() && a2==pairMapPtr[i].getRight()){
			return;

		} 
	}
		
		pairMapPtr[pairMapIndex]=pairClass<T>(a1,a2,a3);
		
		pairMapIndex+=1;


}

template <class T>

int BigramDyn<T>::numGrams(){///////////////////////////
    
    
    return pairIndex;///////////////////
     
}



template <class T>
int BigramDyn<T>::numOfGrams(T p1,T p2){
    pairClass<T> key(p1,p2,0);///YENI paırCLASS OBJESI KEY
  
   
    int count=0;
    
    
    for(int i=0;i<pairIndex;++i)///BUTUN BIGRAMLARIN SAYISI
    {
         
         if((pairAllPtr[i].getLeft()==key.getLeft()) &&////////////////KEY OBJESI BUTUN BIGRAM DEGERLERI ILE KARSILASTIRILIR
				(pairAllPtr[i].getRight()==key.getRight()))
         {
             count+=1;
         }
    }
    
    return count;
   
    
}




template <class T>
pair<T,T> BigramDyn<T>::maxGrams(){
    pair<T,T> maxpair;
    int max;
    
    
    
    for(int i=0;i<pairMapIndex;++i)///////paırMapPtr ARRAYI DNGUYE GIRER
    {   pairMapPtr[i].setValue(0);
        for(int j=0;j<pairIndex;++j)////BUTUN BIGRAMLARIN TUTULDUGU ARRAY DONGUYE GIRER
        {
         
            if((pairMapPtr[i].getLeft()==pairAllPtr[j].getLeft()) && ////////paırMapPtr ve paırAllPtr KARSILASTIRILIR
					(pairMapPtr[i].getRight()==pairAllPtr[j].getRight()))
            {
                pairMapPtr[i].setValue(pairMapPtr[i].getValue()+1);////SADECE TEK TUR BIGRAM TUTULAN ARRAYDEKI
///////////////////////////////////////////////////////////////////////OBJELERIN VALUE KISMI ARTIRILIR
            }
        }
    }
	
    max=pairMapPtr[0].getValue();
	maxpair.first=pairMapPtr[0].getLeft();
	maxpair.second=pairMapPtr[0].getRight();
	
    for(int i=1;i<pairMapIndex;++i)//////paırMapPtr VALUE KISIMLARI KARSILASTIRILIR
    {
        if(max<pairMapPtr[i].getValue())///VALUE KISMI BIGRAMDAN KAC TANE OLDUGUNU TUTAR
        {
            max=pairMapPtr[i].getValue();
            maxpair.first=pairMapPtr[i].getLeft();
			maxpair.second=pairMapPtr[i].getRight();
        }
    }
    return maxpair;
    
    
}



template <class T>

void BigramDyn<T>::print(ostream& output){
	 
    for(int i=0;i<pairMapIndex;++i)////////////////MAP TUTAN ARRAY
    {   pairMapPtr[i].setValue(0);
        for(int j=0;j<pairIndex;++j)///BUTUN BIGRAMLARI TUTAN ARRAY
        {
         
            if((pairMapPtr[i].getLeft()==pairAllPtr[j].getLeft()) && (pairMapPtr[i].getRight()==pairAllPtr[j].getRight()))
            {/////////////////////////////////////////////////////////EGER MAP TUTAN ARRAYDEKI BIGRAMDAN VAR ISE
                pairMapPtr[i].setValue(pairMapPtr[i].getValue()+1);////MAP TUTAN ARRAYDEKI VALUE KISMI ARTTIRILIR
            }
        }
    }
	
	for(int i=1;i<pairMapIndex;++i){///////////////////BUBBLE SORT
		for(int j=0;j<pairMapIndex-i;++j){
			
			if(pairMapPtr[j].getValue()<pairMapPtr[j+1].getValue()){
				
				pairClass<T> temp;
				temp=pairMapPtr[j];
				pairMapPtr[j]=pairMapPtr[j+1];
				pairMapPtr[j+1]=temp;
			}
		}
	}
			
	for(int i=0;i<pairMapIndex;++i){///DUZENLENMIS PAIRCLASS OBJELERININ BASKISI
	
		
		output<<" ( "<<pairMapPtr[i].getLeft()<<" , "<<pairMapPtr[i].getRight()<<" )"<<" = "<<pairMapPtr[i].getValue()<<endl;
		
	}
	
	
}













